utils/protoproxy.go

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2023 LiveKit, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package utils
+
+import (
+	"sync"
+	"time"
+
+	"github.com/frostbyte73/core"
+	"google.golang.org/protobuf/proto"
+)
+
+// ProtoProxy is a caching proxy for protobuf messages that may be expensive
+// to compute. It is used to avoid unnecessary re-generation of Protobufs
+type ProtoProxy[T proto.Message] struct {
+	message         T
+	updateFn        func() T
+	fuse            core.Fuse
+	updateChan      chan struct{}
+	done            chan struct{}
+	dirty           bool
+	refreshedAt     time.Time
+	refreshInterval time.Duration
+	lock            sync.RWMutex
+}
+
+// NewProtoProxy creates a new ProtoProxy that regenerates underlying values at a cadence of refreshInterval
+// this should be used for updates that should be sent periodically, but does not have the urgency of immediate delivery
+// updateFn should provide computations required to generate the protobuf
+// if refreshInterval is 0, then proxy will only update on MarkDirty(true)
+func NewProtoProxy[T proto.Message](refreshInterval time.Duration, updateFn func() T) *ProtoProxy[T] {
+	p := &ProtoProxy[T]{
+		updateChan:      make(chan struct{}, 1),
+		updateFn:        updateFn,
+		done:            make(chan struct{}),
+		fuse:            core.NewFuse(),
+		refreshInterval: refreshInterval,
+	}
+	p.performUpdate(true)
+	if refreshInterval > 0 {
+		go p.worker()
+	}
+	return p
+}
+
+func (p *ProtoProxy[T]) MarkDirty(immediate bool) {
+	p.lock.Lock()
+	p.dirty = true
+	shouldUpdate := immediate || time.Since(p.refreshedAt) > p.refreshInterval
+	p.lock.Unlock()
+	if shouldUpdate {
+		go p.performUpdate(false)
```

Comment:
Could this make things out of order? Looks like it can. If two immediately race, it is possible that the order may be flipped and the older message gets latched?

Reply:
instead of performing the update immediately it could be enqueued on the worker?

Reply:
good point.. let me queue it in the main thread instead.